libbeat/publisher/pipeline/doc.go

```diff
@@ -0,0 +1,53 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/*
+Package pipeline implements event publishing pipelines used in libbeat. These pipelines
+are used to publish events from a source (e.g. an input in Filebeat) all the way to an output
+(e.g. Elasticsearch), via processors and a queue. A pipeline encapsulates all these components
+and abstracts over some key internal components as well to make event publishing possible.
+
+The main component, of course, is a *pipeline* itself. When a new pipeline is created
+a *client* is returned. Users of the pipeline can use this client to publish
+events to the pipeline.
```

Comment:
The pipeline is the "global" instance. We have 2 active pipelines in libbeat (one for event publishing, and one metrics). Entities that want to publish events use `Connect(With)` in order to get a `Client` instance that can be used for publishing. Each `Client` instance is independent. Processors are created and run within the context of the local `Client` and finally push events onto the queue.